### ISSUE TITLE
Re-thinking: Layering of configuration

### ISSUE THREAD

**Issue by matthiasbeyer** [mehcode/config-rs#324]
We currently (as of 0.13.x) support _layering_ of configuration objects to merge configuration values.

This should be supported in the future as well.

---

## Re-Thinking the layering mechanism

We might re-think the idea of layering a bit here.
Right now (AFAICR) the layering is executed on _loading_ of the configuration. In future, we might implement this as an operation that happens on _accessing_ the values.
This way, we would always keep all values from the sources and only "merge" the values when accessing a certain element.

## Example:

```toml
# 1.toml
foo = 1
bar = 2
```

```toml
# 2.toml
bar = 3
```

```
// pseudo code:
Build config object
Register "1.toml"
Register "2.toml"
Access config at "bar"
  - Mechanism finds out that "bar" is from "2.toml" and its value is 3
```

**Comment by szarykott** [comment]
As much as it would be great for debugging purposes, it might impair performance of the solution. Is this proposal motivated by some real life problems that is solves?

**Comment by matthiasbeyer** [comment]
> As much as it would be great for debugging purposes, it might impair performance of the solution. Is this proposal motivated by some real life problems that is solves?

No, I am just thinking from the ground up how this crate could look like in the future.

**Comment by martinvonz** [comment]
Another benefit of preserving the layering after load-time is that the client application can iterate over the layers and print the configs in each layer. That came up in https://github.com/martinvonz/jj/pull/917#discussion_r1057305694
